include/net/nrf_cloud_codec.h

```diff
@@ -20,6 +26,485 @@ extern "C" {
  * @{
  */
 
+/** @brief Object types for encoding and decoding */
+enum nrf_cloud_obj_type {
+	NRF_CLOUD_OBJ_TYPE__UNDEFINED,
+
+	NRF_CLOUD_OBJ_TYPE_JSON,
+
+	NRF_CLOUD_OBJ_TYPE__LAST,
+};
+
+/** @brief Encoded data sources */
+enum nrf_cloud_enc_src {
+	/** Encoded data is not set */
+	NRF_CLOUD_ENC_SRC_NONE,
+	/** Encoded data was set by @ref nrf_cloud_obj_cloud_encode */
+	NRF_CLOUD_ENC_SRC_CLOUD_ENCODED,
+	/** Encoded data was set manually or by @ref NRF_CLOUD_OBJ_PRE_ENC_DEFINE */
+	NRF_CLOUD_ENC_SRC_PRE_ENCODED,
+};
+
+/** @brief Object used for building nRF Cloud messages. */
+struct nrf_cloud_obj {
+
+	enum nrf_cloud_obj_type type;
+	union {
+		cJSON *json;
+	};
+
+	/** Source of encoded data */
+	enum nrf_cloud_enc_src enc_src;
+	/** Encoded data */
+	struct nrf_cloud_data encoded_data;
+};
+
+/** @brief Define an nRF Cloud JSON object.
+ *
+ * This macro defines a codec object with the type of NRF_CLOUD_OBJ_TYPE_JSON.
+ *
+ * @param _name	Name of the object.
+ */
+#define NRF_CLOUD_OBJ_JSON_DEFINE(_name) \
+	struct nrf_cloud_obj _name = { .type = NRF_CLOUD_OBJ_TYPE_JSON, .json = NULL, \
+				       .enc_src = NRF_CLOUD_ENC_SRC_NONE, \
+				       .encoded_data = { .ptr = NULL, .len = 0 } }
```

Comment:
I think the one thing I would consider changing is having this as a macro -- Could these default values be moved into the various `_init` functions? Then users would just declare `struct nrf_cloud_obj name` and call init on the empty/invalid struct.

Maybe this makes global/static declarations of nrf_cloud_obj tough, though. But do we intend to support that?

Or maybe I am being thick-headed and there is some other reason this is a cleaner method. Or maybe my preference against using macros for this sort of thing is just me being weird. What is your stance on the matter?

Pinging @plskeggs as well for his stance

Reply:
https://github.com/jayteemo/fw-nrfconnect-nrf/pull/7#discussion_r1177115225

Reply:
Oh my bad, we've had this conversation before

Reply:
Feel free to resolve, sorry about my forgetfulness

I'm half-tempted to suggest pointing out that its a convenience function in the macro description, but I'm not sure its worth it. Your call